### PR TITLE
Remote sessions support for SyncPlay

### PR DESCRIPTION
**Pull request by OancaAndrei**
**Changes**
Allows a session to enable SyncPlay on another session (only if the feature is supported).

`jellyfin-web` PR [here](https://github.com/jellyfin/jellyfin-web/pull/2203).
`jellyfin-apiclient-javascript` PR [here](https://github.com/jellyfin/jellyfin-apiclient-javascript/pull/94).

**Issues**
Related to #3855.

### CODE REVIEW

File: Jellyfin.Api/Models/SessionDtos/ClientCapabilitiesDto.cs [hunk 1]
```diff
@@ -64,6 +64,11 @@ public class ClientCapabilitiesDto
         /// </summary>
         public string? IconUrl { get; set; }
 
+        /// <summary>
+        /// Gets or sets a value indicating whether session supports SyncPlay.
+        /// </summary>
+        public bool SupportsSyncPlay { get; set; }
```
Comment (crobibero) [inline Jellyfin.Api/Models/SessionDtos/ClientCapabilitiesDto.cs:70]: I think this needs to be nullable for proper backwards compatibility

File: Jellyfin.Api/Controllers/SyncPlayController.cs [hunk 2]
```diff
@@ -73,24 +73,58 @@ public class SyncPlayController : BaseJellyfinApiController
...
+            var joiningSession = RequestHelpers.GetSession(_sessionManager, _authorizationContext, Request);
+
+            // TODO: check if session can control remote session.
+            if (requestData.RemoteSessionId != null)
+            {
+                var remoteSession = _sessionManager.GetSession(requestData.RemoteSessionId);
+                if (!remoteSession.Capabilities.SupportsSyncPlay)
+                {
+                    // Remote session does not support SyncPlay.
+                    return NoContent();
```
Comment (crobibero) [inline Jellyfin.Api/Controllers/SyncPlayController.cs:85]: Should this be `BadRequest` so the client knows it didn't do anything?

Comment (OancaAndrei) [reply Jellyfin.Api/Controllers/SyncPlayController.cs:85]: That's right. Which reminds me that this needs to be handled client side as well.

File: Jellyfin.Api/Controllers/SyncPlayController.cs [hunk 3]
```diff
@@ -66,31 +66,63 @@ public class SyncPlayController : BaseJellyfinApiController
         /// </summary>
         /// <param name="requestData">The group to join.</param>
         /// <response code="204">Group join successful.</response>
+        /// <response code="400">Session does not support SyncPlay.</response>
         /// <returns>A <see cref="NoContentResult"/> indicating success.</returns>
         [HttpPost("Join")]
         [ProducesResponseType(StatusCodes.Status204NoContent)]
```
Comment (crobibero) [inline Jellyfin.Api/Controllers/SyncPlayController.cs:72]:
```suggestion
        [ProducesResponseType(StatusCodes.Status204NoContent)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
```

File: Jellyfin.Api/Controllers/SyncPlayController.cs [hunk 4]
```diff
@@ -66,31 +66,63 @@ public class SyncPlayController : BaseJellyfinApiController
...
 
         /// <summary>
         /// Leave the joined SyncPlay group.
         /// </summary>
+        /// <param name="requestData">The group to leave.</param>
         /// <response code="204">Group leave successful.</response>
+        /// <response code="400">Session does not support SyncPlay.</response>
         /// <returns>A <see cref="NoContentResult"/> indicating success.</returns>
         [HttpPost("Leave")]
         [ProducesResponseType(StatusCodes.Status204NoContent)]
```
Comment (crobibero) [inline Jellyfin.Api/Controllers/SyncPlayController.cs:105]:
```suggestion
        [ProducesResponseType(StatusCodes.Status204NoContent)]
        [ProducesResponseType(StatusCodes.Status400BadRequest)]
```

**Review by crobibero — APPROVED** [review]

File: Emby.Server.Implementations/Session/SessionManager.cs [hunk 5]
```diff
@@ -993,17 +993,9 @@ private bool OnPlaybackStopped(User user, BaseItem item, long? positionTicks, bo
...
-        /// <param name="sessionId">The session identifier.</param>
-        /// <param name="throwOnMissing">if set to <c>true</c> [throw on missing].</param>
-        /// <returns>SessionInfo.</returns>
-        /// <exception cref="ResourceNotFoundException">
-        /// No session with an Id equal to <c>sessionId</c> was found
-        /// and <c>throwOnMissing</c> is <c>true</c>.
-        /// </exception>
-        private SessionInfo GetSession(string sessionId, bool throwOnMissing = true)
+        /// <inheritdoc />
+
```
Comment (Ullmie02) [inline Emby.Server.Implementations/Session/SessionManager.cs:997]:
```suggestion

```

**Comment by dkanada** [comment]
@OancaAndrei did you plan on resolving the TODOs for this one? We can probably fit this feature into the next update but I don't think the group visibility one will make it since it's still a draft.

**Comment by OancaAndrei** [comment]
Hi @dkanada, soo, after a quick look, I think that only these two should be addressed:
> // TODO: check if session can control remote session.

and
> // TODO: reject or allow remote sessions that don't have SyncPlay access?

whereas this other one I think I've left it there open for further discussion (it's an extra check that shouldn't be doing any harm):
> // TODO: is this really needed? These sessions shouldn't join in the first place.

I'll need to setup my dev environment so that I can have a look at the PR again :+1:

**Comment by OancaAndrei** [comment]
Code almost works. The remote session control part looks fine; there are issues when trying to leave a group:
`POST http://localhost:8096/SyncPlay/Leave` -> `415 Unsupported Media Type`, which makes no sense (hopefully not for too long).

**Comment by Bond-009** [comment]
@OancaAndrei Any updates?

**Pull request closed without merging**